src/main/java/com/example/pokebattlez/controller/service/ChallengeService.java

```diff
@@ -0,0 +1,103 @@
+import com.example.pokebattlez.model.entity.Pokemon;
+import com.example.pokebattlez.model.entity.Team;
+import com.example.pokebattlez.model.request.User;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.messaging.simp.SimpMessagingTemplate;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+public class ChallengeService {
+
+    private final BattleService battleService;
+    private final SimpMessagingTemplate template;
+    private final OnlineUsers onlineUsers;
+    private final TeamRepository teamRepository;
+
+    private final List<Challenge> challenges = new ArrayList<>();
+
+    public void register(Long challenger, Long ... challenged) {
+        challenges.add(new Challenge(challenger, List.of(challenged)));
+    }
+
+    public boolean hasActiveChallenge(Long user) {
+        return challenges.stream().anyMatch(challenge -> challenge.challenger.equals(user));
+    }
+
+    public void accept(Long user, Long from) {
+        challenges.stream().filter(challenge -> challenge.challenger.equals(from)).findFirst().ifPresent(challenge -> {
+            challenge.accepts.add(user);
+            if (challenge.isAccepted()) {
+                challenges.remove(challenge);
+                startBattle(challenge);
+            }
+        });
+    }
+
+    public void decline(Long from) {
+        challenges.stream().filter(challenge -> challenge.challenger.equals(from)).findFirst()
+                .ifPresent(challenges::remove);
+    }
+
+    private void startBattle(Challenge challenge) { // TODO Check if all trainers are still online
+        String id = battleService.generateBattle(challenge.challenger, challenge.challenged.toArray(new Long[0]));
+
+        List<String> connections = challenge.challenged.stream().map(challenged1 -> onlineUsers.getConId(challenged1).orElse(null)).collect(Collectors.toList());
+        connections.add(onlineUsers.getConId(challenge.challenger).orElse(null));
+
+        BattleData battleData = new BattleData(id);
+        connections.stream().filter(Objects::nonNull).forEach(connection ->
+                onlineUsers.getUser(connection).ifPresent(user -> {
+                    battleData.trainers.add(user.getId());
+                    battleData.trainerNames.put(user.getId(), user.getName());
+                    teamRepository.findTeamByTrainer_Id(user.getId()).stream()
+                            .peek(team -> battleData.trainerTeams.put(user.getId(), team.getId()))
+                            .map(Team::getPokemon).forEach(team ->
+                                    battleData.availablePokemon.put(user.getId(), team.stream()
+                                            .map(Pokemon::getPosition)
+                                            .collect(Collectors.toList())
```

Comment:
In general I like how you leverage streams to your advantage. But here it's just not clear what's going on.  
At least two methods could be aggregated into their individual representation/implementations.